feature-pitches/002-file-based-workflows/002-file-based-workflows.md

```diff
@@ -0,0 +1,109 @@
+
+As long as Renku workflows are only available on the CLI, only the most advanced users will use them. But Renku’s mission is to make reproducible research *easy* and for *everyone*, not only the computationally advanced! So, our first motivating question is *what can we do to make workflows accessible to a broader audience?*
+
+Additionally, there is a common challenge we run into when introducing Renku workflows to users who *do* want to use workflows: it is cumbersome/just-not-tenable to create and develop (modify) workflows of more than a few steps. Many workflow users have 10+ steps in their workflows [[source](https://sdsc.atlassian.net/wiki/spaces/RENKU/pages/2246541341/Reproducibility+on+HPC)]. Entering 10+ steps on the command line is tedious to do once, but even more significant is that modifying that workflow- which users do all the time as they develop their pipeline- requires recreating and rewiring steps again and again on the CLI. Some users end up writing a script to call a series of Renku run commands [[ORDES example](https://renkulab.io/projects/renku-stories/digirhythm/files/blob/src/build_workflow.sh)]. Renku should provide a built-in way to create and develop multi-step workflows via a file!
+
+## 🍴 Appetite
+
+We’d like to invest 2 sprints (6 weeks) for making workflows more accessible to a broader audience.
+
+## 🎯 Solution
+
+We’d like to offer the ability to define Renku workflows in a file, where the full workflow is laid out for the user to plainly see and edit.
+
+I’ve done a [survey of popular workflow tools](https://sdsc.atlassian.net/wiki/spaces/RENKU/pages/2310701105/Workflow+Tool+Research), and YAML seems to be a common choice for a workflow file, but we leave the final decision to the team.
+
+<img src="fat-marker-file.png" width="400">
+
+The workflow file names a sequence of steps, and each step names its command, inputs, outputs, parameters, etc. Each step should be named so that it can be referenced for running individually. I imagine that the structure of the workflow step definition in the file would be pretty similar to the output of `renku workflow show`.
+
+![renku workflow show](workflow-show.png)
+
+For the scope of this pitch, expressing `renku workflow iterate` in the workflow file definition syntax is _out of scope_, along with any kinds of looping or branching.
+
+### Editing workflow files
+
+The workflow file is opened and modified just like any other code file in the project. This pitch requires no changes to the UI.
+
+The user should be able to create more than one workflow file in a project.
+
+### Running workflow files
+
+The user would still use the CLI to run the workflow, referencing the workflow via the workflow file name. To improve usability for non-CLI users, in tutorials, we could perhaps demonstrate running shell commands from inside a notebook (`!renku workflow execute workflow.yml`).
```

Comment:
I see that it makes sense to reference the workflow name as defined in the workflow file, since non-file-based (CLI) workflows are also referenced by name. But do we also open a small can of worms here, by introducing a need for within-workflow file search? For example, I think it should be a must-have that the user can define more than 1 workflow file within a project. Therefore, when referencing workflow names, we have to check for duplicate workflow names within and between files, and raise error messages appropriately. Perhaps this is a simple enough problem to solve, but just want to surface it before committing to a solution. What do you think?

Reply:
yeah, plus if you see the workflow only getting added to the project metadata _after_ it has been run at least once, it won't be possible to find it by name anyway. I suppose it should be easy enough to check whether the `name` in `renku workflow execute <name>` is a _filename_ or a workflow name that already exists in the project.

Reply:
I'll expand the pitch to say via the workflow _via the workflow file name **or** the workflow name (TBD exactly how this works)_, and leave this up to the implementation team.